### PR TITLE
IGNITE-7024

### PR DESCRIPTION
**Pull request by NSAmelchev**
Implementation of network compression.

- Added new module Compression for compression algorithms implementations.
- Added compression filter to NIO filter chain.
- Added flag to a configuration that enables compression between some nodes. (If flag true at least one of nodes, network traffic between this nodes will be compressed)
- Added compression state flag to sessions interface and to TCP Handshake msg.
- Added JMH benchmark.
- Some changes in next classes to support compression: GridNioServer, TcpCommunicationSPI

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+ */
+
+package org.apache.ignite.internal.util.nio.compression;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.nio.GridNioException;
```
Comment (SharplEr) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java]: Remove unnecessary `import`.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+import org.apache.ignite.internal.util.nio.GridNioSession;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.BUFFER_OVERFLOW;
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.OK;
+
+/**
+ * Class that encapsulate the compression state, compress and decompress logic.
+ */
+public final class GridNioCompressionHandler {
```
Comment (SharplEr) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java]: Decompose into two classes: one thread safe, and one not as a base.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java (hunk 2 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java]: Done.

**Pull request closed without merging**